fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneRecordCursor.java

```diff
@@ -351,23 +375,90 @@ private CompletableFuture<ScoreDocIndexEntry> buildIndexEntryFromScoreDocAsync(@
+                for (Term term : termArray) {
+                    map.putIfAbsent(term.field(), new HashSet<>());
+                    map.get(term.field()).add(term.text().toLowerCase(Locale.ROOT));
+                }
+            }
+        } else if (query instanceof BoostQuery) {
+            BoostQuery boostQuery = (BoostQuery) query;
+            getTerms(boostQuery.getQuery(), map);
+        } else if (query instanceof SynonymQuery) {
+            SynonymQuery synonymQuery = (SynonymQuery) query;
+            for (Term term : synonymQuery.getTerms()) {
+                map.putIfAbsent(term.field(), new HashSet<>());
+                map.get(term.field()).add(term.text().toLowerCase(Locale.ROOT));
+            }
+        } else {
+            throw new RecordCoreException("This lucene query is not supported for highlighting");
+        }
+    }
+
     protected static final class ScoreDocIndexEntry extends IndexEntry {
         private final ScoreDoc scoreDoc;
 
+        private final Map<String, Set<String>> termMap;
+        private final LuceneAnalyzerCombinationProvider analyzerSelector;
+        private final LuceneScanQueryParameters.LuceneQueryHighlightParameters luceneQueryHighlightParameters;
+        private final KeyExpression indexKey;
+
         public ScoreDoc getScoreDoc() {
             return scoreDoc;
         }
 
-        private ScoreDocIndexEntry(@Nonnull ScoreDoc scoreDoc, @Nonnull Index index, @Nonnull Tuple key) {
+        private ScoreDocIndexEntry(@Nonnull ScoreDoc scoreDoc, @Nonnull Index index, @Nonnull Tuple key,
+                                   @Nonnull LuceneScanQueryParameters.LuceneQueryHighlightParameters luceneQueryHighlightParameters, @Nonnull Query query,
+                                   @Nonnull LuceneAnalyzerCombinationProvider analyzerSelector) {
             super(index, key, TupleHelpers.EMPTY);
             this.scoreDoc = scoreDoc;
+            this.luceneQueryHighlightParameters = luceneQueryHighlightParameters;
+            this.termMap = new HashMap<>();
+            this.analyzerSelector = analyzerSelector;
+            this.indexKey = index.getRootExpression();
+            if (luceneQueryHighlightParameters.isHighlight()) {
+                getTerms(query, this.termMap);
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        @Nonnull
+        public <M extends Message> FDBStoredRecord<M> rewriteStoredRecord(@Nonnull FDBStoredRecord<M> record) {
+            if (!luceneQueryHighlightParameters.isHighlight()) {
+                return super.rewriteStoredRecord(record);
+            }
+            M message = record.getRecord();
+            M.Builder builder = message.toBuilder();
+            LuceneDocumentFromRecord.highlightTermsInMessage(indexKey, builder, termMap, analyzerSelector, luceneQueryHighlightParameters);
+            return FDBStoredRecord.newBuilder(record).setRecord((M) builder.build()).build();
```

Comment:
Since it is somewhat awkward to splice this new code into `LuceneIndexQueryPlan`, I suggest, that, rather than trying to push it down like this, pull it up. That is, concretely, add a new plan operator that takes the `FDBQueriedRecord`s that will contain the original stored record and some highlighting information and returns a new stream of rewritten records. Have `LucenePlanner` tack this on in `planLucene`.